File: 3-hard/00041-first-missing-positive/solution.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();

        for (int i{0}; i < n; ++i) {
            if (nums[i] <= 0) {
                nums[i] = INT_MAX;
            }
        }

        for (int i{0}; i < n; ++i) {
            int num = abs(nums[i]) - 1;
            if (num < n) {
                if (nums[num] > 0) {
                    nums[num] *= -1;
                }
            }
        }

        for (int i{0}; i < n; ++i) {
            if (nums[i] > 0) {
                return i + 1;
            }
        }

        return n + 1;
    }
};

File: 3-hard/00041-first-missing-positive/solution.rs
impl Solution {
    pub fn first_missing_positive(mut nums: Vec<i32>) -> i32 {

        let n = nums.len();

        for i in 0..n {
            if nums[i] <= 0 {
                nums[i] = i32::MAX;
            }
        }

        for i in 0..n {
            let num = (nums[i].abs() - 1) as usize;
            if num < n {
                if nums[num] > 0 {
                    nums[num] *= -1;
                }
            }
        }

        for i in 0..n {
            if nums[i] > 0{
                return i as i32 + 1;
            }
        }

        n as i32 + 1
    }
}
